service/controller/v3/resource/instance/create.go

```diff
@@ -9,116 +9,491 @@ import (
+
+	return nil
+}
+
+func containsInstanceVersion(list []compute.VirtualMachineScaleSetVM, version string) bool {
+	for _, v := range list {
+		if *v.InstanceID == version {
+			return true
+		}
+	}
+
+	return false
+}
+
+// findActionableInstance either returns an instance to update or an instance to
+// reimage, but never both at the same time.
+func findActionableInstance(customObject providerv1alpha1.AzureConfig, list []compute.VirtualMachineScaleSetVM, versionValue map[string]string) (*compute.VirtualMachineScaleSetVM, *compute.VirtualMachineScaleSetVM, error) {
+	var err error
 
-		c, err := r.getScaleSetsClient()
+	instanceInProgress := firstInstanceInProgress(customObject, list)
+	if instanceInProgress != nil {
+		return nil, nil, nil
+	}
+
+	var instanceToUpdate *compute.VirtualMachineScaleSetVM
+	if instanceInProgress == nil {
+		instanceToUpdate = firstInstanceToUpdate(customObject, list)
+	}
+
+	var instanceToReimage *compute.VirtualMachineScaleSetVM
+	if instanceToUpdate == nil {
+		instanceToReimage, err = firstInstanceToReimage(customObject, list, versionValue)
 		if err != nil {
-			return microerror.Mask(err)
+			return nil, nil, microerror.Mask(err)
 		}
+	}
 
-		g := key.ResourceGroupName(customObject)
-		s := deploymentNameFunc(customObject)
-		ids := compute.VirtualMachineScaleSetVMInstanceRequiredIDs{
-			InstanceIds: to.StringSlicePtr([]string{
-				instanceID,
-			}),
+	return instanceToUpdate, instanceToReimage, nil
+}
+
+// firstInstanceInProgress returns the first instance in the list not having a
+// final state. In case all instances are in a final state
+// firstInstanceInProgress returns nil.
+func firstInstanceInProgress(customObject providerv1alpha1.AzureConfig, list []compute.VirtualMachineScaleSetVM) *compute.VirtualMachineScaleSetVM {
+	for _, v := range list {
+		if v.ProvisioningState == nil || key.IsFinalProvisioningState(*v.ProvisioningState) {
+			continue
 		}
-		_, err = c.UpdateInstances(ctx, g, s, ids)
-		if err != nil {
-			return microerror.Mask(err)
+
+		return &v
+	}
+
+	return nil
+}
+
+// firstInstanceToReimage returns the first instance to be reimaged. The
+// decision of reimaging an instance is done by comparing the desired version
+// bundle version of the custom object and the current version bundle version of
+// the instance's tags applied. In case all instances are reimaged
+// firstInstanceToReimage return nil.
+func firstInstanceToReimage(customObject providerv1alpha1.AzureConfig, list []compute.VirtualMachineScaleSetVM, versionValue map[string]string) (*compute.VirtualMachineScaleSetVM, error) {
+	if versionValue == nil {
+		return nil, microerror.Mask(versionBlobEmptyError)
+	}
+
+	for _, v := range list {
+		desiredVersion := key.VersionBundleVersion(customObject)
+		instanceVersion, ok := versionValue[*v.InstanceID]
+		if !ok {
+			return nil, microerror.Mask(versionBlobEmptyError)
```

Comment:
I don't follow at all here. This is the way how it works. When we continue the logic breaks. It is also that all this is not really cool anyway and I want to ease the version magic once we have the version information in the guest cluster's k8s nodes. This will be refactored anyway and I would just not invest more time here now.